sw-sysemu/memory/etsoc1/pma.cpp

```diff
@@ -0,0 +1,726 @@
+/*-------------------------------------------------------------------------
+* Copyright (c) 2025 Ainekko, Co.
+* SPDX-License-Identifier: Apache-2.0
+*-------------------------------------------------------------------------*/
+
+#include <cinttypes>
+
+#include "memory/pma.h"
+#include "memmap.h"
+#include "emu_gio.h"
+#include "system.h"
+#include "utility.h"
+#include "insn_util.h"
+#ifdef SYS_EMU
+#include "sys_emu.h"
+#include "checkers/mem_checker.h"
+#endif
+
+#ifdef SYS_EMU
+#define SYS_EMU_PTR cpu.chip->emu()
+#endif
+
+namespace bemu {
+
+//------------------------------------------------------------------------------
+// Exceptions
+
+static inline Privilege effective_execution_mode(const Hart& cpu, mem_access_type macc)
+{
+    // Read mstatus
+    const uint64_t  mstatus = cpu.mstatus;
+    const int       mprv    = (mstatus >> MSTATUS_MPRV) & 0x1;
+    const Privilege mpp     = Privilege((mstatus >> MSTATUS_MPP ) & 0x3);
+    const Privilege prv     = cpu.prv;
+    return (macc == Mem_Access_Fetch) ? prv : (mprv ? mpp : prv);
+}
+
+
+//------------------------------------------------------------------------------
+// PMA checks
+
+#define MPROT_IO_ACCESS_MODE(x)    ((x) & 0x3)
+#define MPROT_DISABLE_PCIE_ACCESS  0x04
+#define MPROT_DISABLE_OSBOX_ACCESS 0x08
+#define MPROT_DRAM_SIZE_8G         0x00
+#define MPROT_DRAM_SIZE_16G        0x10
+#define MPROT_DRAM_SIZE_24G        0x20
+#define MPROT_DRAM_SIZE_32G        0x30
+#define MPROT_DRAM_SIZE(x)         ((x) & 0x30)
+#define MPROT_ENABLE_SECURE_MEMORY 0x40
+
+#define PP(x)   (int(((x) & ESR_REGION_PROT_MASK) >> ESR_REGION_PROT_SHIFT))
+
+
+[[noreturn]] static void throw_page_fault(uint64_t addr, mem_access_type macc)
```

Comment:
These are not ETSOC1 dependent, they're part of the MMU and should be exported as part of the MMU as such.

Only the PMA specific things should be here.